composer/callbacks/checkpoint_saver.py

```diff
@@ -39,6 +47,58 @@
 __all__ = ['CheckpointSaver']
 
 
+def _upload_symlink_file(
+    remote_backend_name: str,
+    backend_kwargs: dict[str, Any],
+    remote_symlink_file_name: str,
+    local_symlink_file_name: str,
+    num_attempts: int,
+    remote_checkpoint_file_names: list[str],
+    main_process_pid: int,
+    is_remote_upload_failed: multiprocessing.Event, # pyright: ignore[reportGeneralTypeIssues]
+    max_wait_time_in_seconds: int = 3600,
+    wait_before_next_try_in_seconds: float = 30,
+):
+    """Wait the checkpoint file uploading to finish and start symlink file uploading."""
+    start_time = time.time()
+    object_store = build_remote_backend(remote_backend_name, backend_kwargs)
+
+    for remote_file_name in remote_checkpoint_file_names:
+        while True:
+            if is_remote_upload_failed.is_set():
+                log.debug(f'Stop symlink uploading since the checkpoint files uploading failed')
+                return
+            # Return if parent process exits
+            try:
+                os.kill(main_process_pid, 0)
+            except OSError:
+                return
+            try:
+                object_store.get_object_size(remote_file_name)
+                break
+            except Exception as e:
+                if not isinstance(e, FileNotFoundError):
+                    log.debug(f'Got exception {type(e)}: {str(e)} when accessing remote file {remote_file_name}')
+                time.sleep(wait_before_next_try_in_seconds)
+            if time.time() - start_time > max_wait_time_in_seconds:
+                raise RuntimeError(
+                    f'Checkpoint file {remote_file_name} uploading not finished after {max_wait_time_in_seconds} seconds',
+                )
+
+    log.debug(f'Uploading symlink file {remote_symlink_file_name}')
```

Comment:
Once the symlink file is uploaded, it's guaranteed to be pointing at a complete checkpoint right? How do you handle if the next checkpoint is partially uploaded but the run crashes. If overwrite is False, it'll crash on resumption right?

Reply:
yes, there is a checkpoint v2 task: corrupted checkpoint loading, it will do two things:
1. update the symlink to point to the latest complete checkpoint
2. set the overwrite to true after loading the checkpoint.

Reply:
Is this PR's behavior enabled in this PR? Or will you enable this PR's behavior in the corrupted checkpoint loading PR?

Reply:
Discussed offline, feature will be in production once this is merged